ast_tools/common.py

```diff
@@ -11,47 +11,85 @@
 
 import astor
 
+import libcst as cst
+
 from ast_tools import stack
 from ast_tools.stack import SymbolTable
 from ast_tools.visitors import used_names
 
-__ALL__ = ['exec_in_file', 'exec_def_in_file', 'get_ast', 'gen_free_name']
+__ALL__ = ['exec_in_file', 'exec_def_in_file', 'exec_str_in_file', 'get_ast', 'get_cst', 'gen_free_name']
 
 DefStmt = tp.Union[ast.FunctionDef, ast.AsyncFunctionDef, ast.ClassDef]
 
 def exec_def_in_file(
         tree: DefStmt,
         st: SymbolTable,
         path: tp.Optional[str] = None,
-        file_name: tp.Optional[str] = None) -> None:
+        file_name: tp.Optional[str] = None,
+        serialized_tree: tp.Optional[DefStmt] = None,
+        ):
     """
     execs a definition in a file and returns the definiton
+
+    For explanation of serialized_tree see
+    https://github.com/leonardt/ast_tools/issues/46
     """
     if file_name is None:
         file_name = tree.name + f'{hash(tree)}.py'
 
-    return exec_in_file(tree, st, path, file_name)[tree.name]
+    return exec_in_file(tree, st, path, file_name, serialized_tree)[tree.name]
 
 
 def exec_in_file(
         tree: ast.AST,
         st: SymbolTable,
         path: tp.Optional[str] = None,
-        file_name: tp.Optional[str] = None) -> None:
+        file_name: tp.Optional[str] = None,
+        serialized_tree: tp.Optional[DefStmt] = None,
+        ):
+
+    """
+    execs an ast as a module and returns the modified enviroment
+
+    For explanation of serialized_tree see
+    https://github.com/leonardt/ast_tools/issues/46
+    """
+
+    source = astor.to_source(tree)
+    if serialized_tree is None:
+        serialized_source = source
```

Comment:
Hmm, wondering if we should move line 58 inside the if so we avoid doing it if we have a serialized tree?  Not a big issue but could slightly improve performance to avoid the extra to_source

Reply:
never mind, I just realized source and serialized source are both passed through, ignore the above comment